sockeye/train.py

```diff
@@ -132,278 +139,460 @@ def main():
+                                           vocab_target=vocab_target,
+                                           vocab_source_path=args.source_vocab,
+                                           vocab_target_path=args.target_vocab,
+                                           batch_size=args.batch_size,
+                                           batch_by_words=args.batch_type == C.BATCH_TYPE_WORD,
+                                           batch_num_devices=batch_num_devices,
+                                           fill_up=args.fill_up,
+                                           max_seq_len_source=max_seq_len_source,
+                                           max_seq_len_target=max_seq_len_target,
+                                           bucketing=not args.no_bucketing,
+                                           bucket_width=args.bucket_width)
+
+
+def create_lr_scheduler(args: argparse.Namespace, resume_training: bool,
+                        training_state_dir: str) -> lr_scheduler.LearningRateScheduler:
+    """
+    Create the learning rate scheduler.
+
+    :param args: Arguments as returned by argparse.
+    :param resume_training: Gives if an interrupted training should be resumed.
+    :param training_state_dir: Directory where the training state is stored.
+    :return: The learning rate scheduler.
+    """
+    learning_rate_half_life = none_if_negative(args.learning_rate_half_life)
+    # TODO: The loading for continuation of the scheduler is done separately from the other parts
+    if not resume_training:
+        lr_scheduler_instance = lr_scheduler.get_lr_scheduler(args.learning_rate_scheduler_type,
+                                                              args.checkpoint_frequency,
+                                                              learning_rate_half_life,
+                                                              args.learning_rate_reduce_factor,
+                                                              args.learning_rate_reduce_num_not_improved,
+                                                              args.learning_rate_schedule,
+                                                              args.learning_rate_warmup)
+    else:
+        with open(os.path.join(training_state_dir, C.SCHEDULER_STATE_NAME), "rb") as fp:
+            lr_scheduler_instance = pickle.load(fp)
+    return lr_scheduler_instance
+
+
+def create_encoder_config(args: argparse.Namespace, vocab_source_size: int,
+                          config_conv: Optional[encoder.ConvolutionalEmbeddingConfig]) -> Config:
+    """
+    Create the encoder config.
+
+    :param args: Arguments as returned by argparse.
+    :param vocab_source_size: The source vocabulary.
+    :param config_conv: The config for the convolutional encoder (optional).
+    :return: The encoder config.
+    """
+    encoder_num_layers, _ = args.num_layers
+    config_encoder = Config()  # Make mypy happy
```

Comment:
why not `config_encoder = None  # type: Optional[Config]` ?

Reply:
Because I was not so familiar with mypy :-D Will do!